### DISCUSSION TITLE
TCPPING: Unable to use proxy between KC instances

### DISCUSSION THREAD

**Discussion by gertjankoorn** [keycloak/keycloak#25844] [Q&A]
When trying to configure shared caching with TCPPING in Nomad, i am unable to connect them to each other. 
A little context: The goal is to run multiple instances of Keycloak on our Nomad cluster. However, running them normally in a single task (like a deployment in Kubernetes), they are not able to communicate to each other. This is intended behaviour by Nomad, because an instance of a service should only be reached by the service in front of it. With Kubernetes, it's a possibility to use a headless service. However, this is not an option in Nomad unfortunately.
A workaround i'm trying now is to manually create a task for each Keycloak instance. Each instance has the other instances' Consul Connect upstreams in their TCPPING configuration, which goes through the sidecar. An overview of they connect:
![kc](https://github.com/keycloak/keycloak/assets/29374639/075a2912-2c26-48ef-bc5c-779400d4052c)
The problem here, is that the infinispan traffic goes through the consul connect proxies, which does not seem to work. The port is open and when i do a telnet to the upstream address and port, i do indeed see an error in the logs of keycloak that it doesn't know what to do with telnet traffic. That seems to me the connection `KC1 -> connect proxy KC1 -> connect proxy KC2 -> KC2` is open. I have the feeling it has something to do specifically with the infinispan cache traffic, but cannot find what's different. 

To test this further, i thought to try this in Docker compose with a Nginx proxy between the instances using the `stream` directive. The config looks like this:
```
stream {
    upstream keycloak1 {
        server <ip01>:<port01>;
    }
    upstream keycloak2 {
        server <ip02>:<port02>;
    }
    server {
        listen <proxy_port01>;
        proxy_pass keycloak1;
    }
    server {
        listen <proxy_port02>;
        proxy_pass keycloak2;
    }
}
```

The ispn configuration is use is: 
```
    <jgroups>
        <stack name="tcpping" extends="tcp">
          <TCP external_addr="${KC_BIND_ADDR}" bind_addr="${KC_BIND_ADDR}" bind_port="${TCP_BIND_PORT}" />
          <TCPPING initial_hosts="${INITIAL_HOSTS}"
                   max_dynamic_hosts="0"
                   port_range="0"
                   stack.combine="REPLACE"
                   stack.position="MPING"/>
        </stack>
    </jgroups>

    <cache-container name="keycloak">
        <transport lock-timeout="60000" stack="tcpping"/>
```
With `INITIAL_HOSTS` being the ip:port of the proxy. 
I get the same experience as in Nomad, where i can telnet with the resulting error in the logs, but the KC's don't seem to find the other instance to build a cache pool. To imitate the behaviour of Nomad, i've put both KC instances in a different Docker network, so they cannot use UDP multicast (which is done by default). When i point them to each other directly with `INITIAL_HOSTS`, they do share cache (to confirm the TCPPING config works correctly). 

Does anyone have an idea what might be the issue here? Or a suggestion what to try?

**Comment by gertjankoorn** [comment]
@akbar1214 Suggested i use jbdc_ping. However, jbdc_ping is as far as i know only used for the discovery/creation of the cache cluster. The actual communication (transport protocol) is still direct communication (through the same Consul proxies). So the same issue still exists with either jbdc_ping or tcpping. See [point 3](https://www.keycloak.org/2019/05/keycloak-cluster-setup.html) and 'Available transport stacks' in [the docs](https://www.keycloak.org/server/caching).
Please let me know if i got this wrong.

**Reply by sschu** [reply]
This is also my understanding. To debug this, you could try to use tcpdump to see what is going on on the network layer.

**Reply by gertjankoorn** [reply]
Yes i've tried that on the nginx proxy that i have running in docker-compose (to imitate the consul connect behaviour). I can see the traffic coming from both keycloak instances, and the proxy seems to forward the traffic just fine:
```
15:34:23.589420 eth0  In  IP keycloak2.network.47847 > cache-proxy.9921: Flags [P.], seq 1469:1582, ack 1, win 502, options [nop,nop,TS val 1042133391 ecr 3887654274], length 113
15:34:23.589449 eth0  Out IP cache-proxy.9921 > keycloak2.network.47847: Flags [.], ack 1582, win 509, options [nop,nop,TS val 3887676466 ecr 1042133391], length 0
15:34:23.589726 eth1  Out IP cache-proxy.47858 > keycloak2.network2.9192: Flags [P.], seq 1469:1582, ack 1, win 502, options [nop,nop,TS val 4165868360 ecr 1848259268], length 113
15:34:23.589808 eth1  In  IP keycloak2.network2.9192 > cache-proxy.47858: Flags [.], ack 1582, win 509, options [nop,nop,TS val 1848281461 ecr 4165868360], length 0
15:34:23.589864 eth0  Out IP cache-proxy.50622 > keycloak1.network.9191: Flags [P.], seq 1469:1582, ack 1, win 502, options [nop,nop,TS val 3849126541 ecr 4234014214], length 113
15:34:23.589958 eth0  In  IP keycloak1.network.9191 > cache-proxy.50622: Flags [.], ack 1582, win 509, options [nop,nop,TS val 4234036406 ecr 3849126541], length 0
15:34:37.030508 eth1  In  IP keycloak1.network2.36879 > cache-proxy.9922: Flags [P.], seq 1456:1568, ack 1, win 502, options [nop,nop,TS val 1322366221 ecr 1850381537], length 112
15:34:37.030558 eth1  Out IP cache-proxy.9922 > keycloak1.network2.36879: Flags [.], ack 1568, win 509, options [nop,nop,TS val 1850411346 ecr 1322366221], length 0
15:34:37.030762 eth0  In  IP keycloak1.network.43505 > cache-proxy.9921: Flags [P.], seq 1456:1568, ack 1, win 502, options [nop,nop,TS val 4234049846 ecr 3849110172], length 112
15:34:37.030781 eth0  Out IP cache-proxy.9921 > keycloak1.network.43505: Flags [.], ack 1568, win 509, options [nop,nop,TS val 3849139982 ecr 4234049846], length 0
15:34:37.030860 eth1  Out IP cache-proxy.35052 > keycloak2.network2.9192: Flags [P.], seq 1456:1568, ack 1, win 502, options [nop,nop,TS val 4165881802 ecr 1848265093], length 112
15:34:37.030977 eth1  In  IP keycloak2.network2.9192 > cache-proxy.35052: Flags [.], ack 1568, win 509, options [nop,nop,TS val 1848294902 ecr 4165881802], length 0
15:34:37.031066 eth0  Out IP cache-proxy.57268 > keycloak1.network.9191: Flags [P.], seq 1456:1568, ack 1, win 502, options [nop,nop,TS val 3849139982 ecr 4234020038], length 112
15:34:37.031141 eth0  In  IP keycloak1.network.9191 > cache-proxy.57268: Flags [.], ack 1568, win 509, options [nop,nop,TS val 4234049847 ecr 3849139982], length 0
```

Keycloak doesn't log anything with the default log level, and on debug it's quite unreadable. What i do see is:
```
TRACE [org.jgroups.protocols.TCP] (keycloak-cache-init) keycloak1-2825: sending msg to <keycloak2-ip>:<keycloak2-port>, src=keycloak1-2825, size=103, headers are TCPPING: [GET_MBRS_REQ cluster=ISPN initial_discovery=true], TP: [cluster=ISPN]
TRACE [org.jgroups.protocols.TCP] (TQ-Bundler-7,keycloak1-2825) <keycloak1-ip>:<keycloak1-port>: connecting to <keycloak2-ip>:<keycloak2-port>
TRACE [org.jgroups.protocols.TCP] (Connection.Receiver [<keycloak1-ip>:37995 - <keycloak2-ip>:<keycloak2-port>]-11,keycloak1-2825) <keycloak1-ip>:<keycloak1-port>: removed connection to <keycloak2-ip>:<keycloak2-port>
```
Unfortunately nothing about something like "connection refused", or "i don't like the traffic i receive from the proxy".

**Reply by sschu** [reply]
Does your proxy actually terminate the TCP connections and create a new one to the target? It could be that this just doesn't work for the internal Infinispan protocol. They might set some socket options that are not compatible, something in that direction.

**Reply by gertjankoorn** [reply]
> Does your proxy actually terminate the TCP connections and create a new one to the target?

As far as i know, yes it does. I'm not sure how to debug that however, or the socket options compatibility. Do you perhaps have a suggestion?

**Reply by sschu** [reply]
I would assume `netstat` should be able to show them, but not sure.

**Comment by kochen** [comment]
@gertjankoorn were you able to resolve this and make multiple keycloak instance cluster in Nomad?

**Reply by gertjankoorn** [reply]
@kochen yes! It turns out i had the right solution after all, except my testing case wasn't right. At first i tested on Nomad itself, but the cycle took to long to test every time. I tried to replicate the setup by using a reverse proxy for the cache traffic in Docker compose, but for some reason the traffic wouldn't go trough it. When i installed Consul Connect on Docker compose, it worked like a charm. We did change to jbdc ping, now that the traffic between the instances works, but as far as i remember both solutions worked. The config we use now is: 
```
    <jgroups>
        <stack name="jdbc-ping-tcp" extends="tcp">
        	<TCP bind_addr="${env.KC_BIND_ADDR}" bind_port="${env.KC_BIND_PORT}"/>
            <JDBC_PING connection_driver="com.mysql.cj.jdbc.Driver"
                       connection_username="${env.KC_DB_USERNAME}"
                       connection_password="${env.KC_DB_PASSWORD}"
                       connection_url="${env.KC_DB_URL}"
                       info_writer_sleep_time="500"
					   remove_all_data_on_view_change="true"
                       stack.combine="REPLACE"
                       stack.position="MPING" />
        </stack>
    </jgroups> 
```
With this in the Dockerfile:
```
ENV KC_CACHE=ispn
ENV KC_CACHE_CONFIG_FILE=cache-ispn-tcp.xml
```

Please let me know if you have any other questions.

**Reply by kochen** [reply]
Thanks @gertjankoorn - I'm also trying to use jdbc ping.

Any chance you could share the important parts of the working nomad job?

**Reply by gertjankoorn** [reply]
Ofcourse. 
The setup of the job is:
```
job {
  group keycloak01 {
    service keycloak01-cache {
       port: 9191
       upstream -> keycloak02-cache (port 9192)
    }
    service keycloak {
      port: 8080
    }
    task keycloak01 {
      KC_BIND_PORT: 9191
      KC_BIND_ADDR: 127.0.0.1
    }
  }
  group keycloak02 {
    service keycloak02-cache {
       port: 9192
       upstream -> keycloak01-cache (port 9191)
    }
    service keycloak {
      port: 8080
    }
    task keycloak02 {
      KC_BIND_PORT: 9192
      KC_BIND_ADDR: 127.0.0.1
    }
  }
}
```
This should be the gist of the job. Not the prettiest, but it works. In this case each instance has it's own group, which is required if you want them to talk to each other. All instances have two services. A service with their own number, which is used by the other services to reach them on. If you want 3 instances, you would have to add another upstream to for all instances. The other service is generic, using the same name. That one is used for ingress traffic, because this creates a "pool" of all instances together.
I hope it makes sense.

**Reply by kochen** [reply]
Thanks!

What is port `9191` in your case? is that the `local_bind_port` (connecting to the service's port `8080`)?

This is going to be hard (if not impossible) to make it dynamic.
I'm hoping to be able to achieve something like this:
```
job {
  group {
    count = 3. // <- easily increased
    service {}
    task { 
      KC_BIND_ADDR = "${attr.unique.network.ip-address}"
      KC_BIND_PORT = "${NOMAD_PORT_bind}"
    }
  }
}
```

**Reply by gertjankoorn** [reply]
It is the `local_bind_port` indeed, but it's not connecting to `8080`. `8080` is only used for ingress/customer traffic. The `9191`/`9192` ports are for the second service that each instance has, used to connect the instances to each other. See the red line in the graph of my initial post (i see some mistakes in the graph, ignore that please :)).

As for making it dynamic, that's a challenge indeed. For us, the only way to connect instances is via Consul Connect upstreams, as DNS isn't open to use and the local IP is only reachable from the host it's on. Since we have multiple hosts, this isn't feasible. 
This way it's not dynamic, but better than having only one instance!

**Reply by kochen** [reply]
is this mapped somehow to port 7800? I thought that was the port needed...

**Reply by gertjankoorn** [reply]
By default Infinispan listens to 7800 indeed, but by providing KC_BIND_PORT to the infinispan config file, this is overridden.
